pinot-spi/src/main/java/org/apache/pinot/spi/accounting/ThreadAccountant.java

```diff
@@ -0,0 +1,72 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.spi.accounting;
+
+import javax.annotation.Nullable;
+
+
+public interface ThreadAccountant {
+
+  /**
+   * clear thread accounting info
+   */
+  void clear();
+
+  /**
+   * check if the corresponding runner thread of current thread is interrupted
+   */
+  boolean isRootThreadInterrupted();
+
+  /**
+   * Task tracking info
+   * @param queryId query id string
+   * @param taskId a unique task id
+   * @param parentContext the parent execution context, null for root(runner) thread
+   */
+  void createExecutionContext(String queryId, int taskId, @Nullable ExecutionContext parentContext);
```

Comment:
Can you elaborate a little bit on the ExecutionContext hierarchy ?

taskId is used to submit tasks for segment level (operator tree level) execution to the worker threads. Each worker thread gets one or more segments that they process sequentially. 

Is the `taskId` in `createExecutionContext` same as above `taskId` implying per query there are as many ExecutionContext as the number of segments executed by that query ?

Reply:
The hierarchy of `ExecutionContext` is implementation dependent as far as the interface concerns ( the context can point to its anchor thread / be independent per thread / have a pointer to its parent ). In our impl it's per thread.